cmd/kubectl-gadget/filetop.go

```diff
@@ -0,0 +1,265 @@
+	RunE: func(cmd *cobra.Command, args []string) error {
+		var err error
+
+		nodeStats = make(map[string][]types.Stats)
+
+		if len(args) == 1 {
+			interval, err = strconv.Atoi(args[0])
+			if err != nil {
+				return fmt.Errorf("%q is not a valid value for interval", args[0])
+			}
+		}
+
+		config := &utils.TraceConfig{
+			GadgetName:       "filetop",
+			Operation:        "start",
+			TraceOutputMode:  "Stream",
+			TraceOutputState: "Started",
+			CommonFlags:      &params,
+			Parameters: map[string]string{
+				"max_rows":  strconv.Itoa(maxrows),
+				"interval":  strconv.Itoa(interval),
+				"sort_by":   sortby_str,
+				"all_files": strconv.FormatBool(allFiles),
+			},
+		}
+
+		startprint()
+
+		if err := utils.RunTraceStreamCallback(config, callback); err != nil {
+			return fmt.Errorf("error running trace: %w", err)
+		}
+
+		return nil
+	},
+	SilenceUsage: true,
+	PreRunE: func(cmd *cobra.Command, args []string) error {
+		var err error
+		sortby, err = types.ParseSortBy(sortby_str)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	},
+	Args: cobra.MaximumNArgs(1),
+}
+
+func init() {
+	filetopCmd.Flags().IntVarP(&maxrows, "maxrows", "r", 20, "Maximum rows to print")
+	filetopCmd.Flags().StringVarP(&sortby_str, "sort", "", "rbytes", "Sort column")
+	filetopCmd.Flags().BoolVarP(&allFiles, "all-files", "a", false, "Include non-regular file types (sockets, FIFOs, etc)")
+
+	rootCmd.AddCommand(filetopCmd)
+	utils.AddCommonFlags(filetopCmd, &params)
+}
+
+func callback(line string, node string) {
+	mu.Lock()
+	defer mu.Unlock()
+
+	var event types.Event
+
+	if err := json.Unmarshal([]byte(line), &event); err != nil {
+		fmt.Fprintf(os.Stderr, "error unmarshalling json: %s", err)
+		return
+	}
+
+	if event.Error != "" {
+		fmt.Fprintf(os.Stderr, "error in node %s:%s", event.Node, event.Error)
+		return
+	}
+
+	nodeStats[node] = event.Stats
+}
+
+func startprint() {
+	ticker := time.NewTicker(time.Duration(interval) * time.Second)
+
+	go func() {
+		for {
+			print()
+			_ = <-ticker.C
+		}
+	}()
+}
+
+func print() {
+	// sort and print events
+	mu.Lock()
+
+	stats := []types.Stats{}
+	for _, stat := range nodeStats {
+		stats = append(stats, stat...)
+	}
+	nodeStats = make(map[string][]types.Stats)
+
+	mu.Unlock()
+
+	types.SortStats(stats, sortby)
+
+	switch params.OutputMode {
+	case utils.OutputModeColumns:
+		if term.IsTerminal(int(os.Stdout.Fd())) {
+			utils.ClearScreen()
+		} else {
+			fmt.Println("")
+		}
+		fmt.Printf("%-16s %-16s %-16s %-16s %-7s %-16s %-6s %-6s %-7s %-7s %1s %s\n",
+			"NODE", "NAMESPACE", "POD", "CONTAINER",
+			"TID", "COMM", "READS", "WRITES", "R_Kb", "W_Kb", "T", "FILE")
+		for idx, event := range stats {
+			if idx == maxrows {
+				break
+			}
+			fmt.Printf("%-16s %-16s %-16s %-16s %-7d %-16s %-6d %-6d %-7d %-7d %c %s\n",
+				event.Node, event.Namespace, event.Pod, event.Container,
+				event.Tid, event.Comm, event.Reads, event.Writes, event.ReadBytes/1024,
```

Comment:
You are printing TID while for other gadgets we print PID.
I am fine with having the TID but we need to harmonize our gadgets so maybe you can add the PID too (the BPF code gets both PID and TID as well as your go code).

Reply:
Good idea. I switched print PID by default but the TID is also available if the user uses costum-columns or json.

Reply:
Cool, thank you!
I am thinking I should maybe write a function like [`print_container_info_header()`](https://github.com/kinvolk/bcc/pull/10/commits/0db55450a6eac55067caa6614588f57e6877a96b#diff-182a3788ae97df7886b4384d7088909a93e7c4e46800f946f1e88b34a9e11b04R142) but for our golang gadgets.

Reply:
I think it's not needed. IMO it adds extra complexity. Perhaps in the future we could have a sort of helpers to print well known fields: namespace, container pod and node names, pid, tid, filepath, etc.